Zend/zend_compile.c

```diff
@@ -5341,7 +5352,13 @@ void zend_compile_foreach(zend_ast *ast) /* {{{ */
 
 	zend_end_loop(opnum_fetch, &reset_node);
 
-	opline = zend_emit_op(NULL, ZEND_FE_FREE, &reset_node, NULL);
+	zend_emit_op(NULL, ZEND_FE_FREE, &reset_node, NULL);
+
+	if (by_ref && value_ast->kind != ZEND_AST_ARRAY && !contains_append(value_ast)) {
+		znode var_node;
+		zend_compile_var(&var_node, value_ast, BP_VAR_UNSET, /* by_ref */ false);
+		zend_emit_op(NULL, ZEND_UNWRAP_REF, &var_node, NULL);
```

Comment:
> This RFC proposes to only perform the reference unwrapping for simple variables of the form $value (“CV”), but not for complex variables.

The implementation seems out of sync with the RFC. Are you planning to forbid `$var[0]`, `$$var`, and so on and change the check to value_ast->kind == ZEND_AST_VAR (and that the variable name is a zval of type string)

The decision to limit it to simple variables makes sense to me, to avoid cases such as the following.

```php
<?php

class CustomArrayAccess implements ArrayAccess {
    public function offsetGet($offset): mixed {
        echo "Called offsetGet\n";
        return new stdClass();
    }
    public function offsetSet($offset, $value): void {
        echo "Called offsetSet\n";
    }
    public function offsetUnset($offset): void {
        echo "Called offsetUnset\n";
    }
    public function offsetExists($offset): bool {
        echo "Called offsetExists\n";
        return true;
    }
}
call_user_func(function () {
    $y = [];
    foreach ([] as &$y[print("call\n")]) {}
    $target = new CustomArrayAccess();
    foreach ([] as &$target[0]) {
        echo "inside\n";
    }
});
```

```
call
Called offsetGet
```

Reply:
Yeah, I implemented this before deciding on the behavior in the RFC. I've now updated the implementation.